### PR TITLE
Add masked_softmax to speed up masking in multihead attention

### PR DESCRIPTION
**Pull request by mickypaganini**
Implemented new `masked_softmax`, which extends `softmax` by allowing for an extra mask tensor argument. Equivalent to `inputs.masked_fill(mask, float('-inf'))` plus a `softmax`, thus, effectively, ignoring the masked entries.

Naive python implementation found at `torch/nn/functional.py`. Implemented a custom C++ kernel for CPU in `aten/src/ATen/native/SoftMax.cpp`. 

Provided tests at `test/test_nn.py` and benchmarks at `benchmarks/operator_benchmark/pt/` by adding the files `masked_softmax_test.py`, `masked_softmax_cpp_test.py`, and `multi_head_attention_forward_test.py`. 

Limitations: inputs must be float32. Mask must be bool or uint8 (cast to bool). Only CPU kernel.

Example usage: 
```python
inputs = torch.rand((2, 2), dtype=torch.float32)
mask = torch.tensor([[1, 1], [0, 1]], dtype=torch.bool)
outputs = torch.nn.functional.masked_softmax(inputs, mask, dim=-1)
```
or
```python
outputs = inputs.masked_softmax(mask, dim=-1)
```

Preliminary benchmarks show promise in speeding up the `forward` of `nn.MultiheadAttention` by replacing the `masked_fill`+ `softmax` with this new `masked_softmax`:

image
The following parameters were used for `nn.MultiheadAttention` in the benchmark above: `L: 256, N: 5, E: 16, S: 256, num_heads: 4, device: cpu`.

TODO: backward computation; documentation.

### CODE REVIEW

File: aten/src/ATen/native/SoftMax.cpp [hunk 1]
```diff
@@ -115,6 +115,62 @@ void host_softmax_backward(
...
+              max_input = std::max(max_input, input_data[d * dim_stride]);
+              unmasked.push_back(d);
+            }
+            else {
+              output_data[d * dim_stride] = 0;
+            }
+          }
+
+          acc_type<scalar_t, false> tmpsum = 0;
+          for (auto d: unmasked) {
```
Comment (cpuhrsch) [inline aten/src/ATen/native/SoftMax.cpp:162]: Did you find a speedup in using this additional std::vector? Otherwise I'd expect repeating the check above and iterating over the entire data again to be faster instead of saving out the indices into a freshly allocated std::vector and then iterating over those simply because whatever portion of "mask_data" you've just used is still available in cache anyway. So you could try 
```
for (int64_t d = 0; d < dim_size; d++){
    if (mask_data[d * dim_stride] == false) {
```
again instead of
```
for (auto d: unmasked) {
```

File: aten/src/ATen/native/SoftMax.cpp [hunk 2]
```diff
@@ -141,6 +197,28 @@ Tensor softmax_cpu(const Tensor& input_, const int64_t dim_, const bool half_to_
   return output;
 }
 
+Tensor masked_softmax_cpu(const Tensor& input_, const Tensor& mask_, const int64_t dim_, const bool half_to_float) {
```
Comment (cpuhrsch) [inline aten/src/ATen/native/SoftMax.cpp:200]: You could add two more checks here.
One to verify that both mask and input are actually on CPU and another to check the dtypes and optionally mask type conversion.

File: test/test_nn.py [hunk 3]
```diff
@@ -3155,6 +3155,139 @@ def test_pad_scalar_error(self):
...
+
+        mask = torch.triu(torch.ones_like(inputs), diagonal=1).to(torch.uint8)
+        with self.assertWarnsRegex(UserWarning, "deprecated"):
+            torch.nn.functional.masked_softmax(inputs, mask, dim=-1)
+
+        mask = torch.rand((2, 2))
+        with self.assertRaises(TypeError):
+            torch.nn.functional.masked_softmax(inputs, mask, dim=-1)
+
+    def test_masked_softmax_cpp_mask_dtype(self):
```
Comment (cpuhrsch) [inline test/test_nn.py:3208]: You could parameterize these tests and do something like this:
```
def _test_masked_softmax(input, mask):
    def reference(input, mask, dim, dtype):
        input = input.masked_fill(mask, float('-inf'))
        return torch.nn.functional.softmax(input, dim=dim, dtype=dtype)
   self.assertEqual(reference(input, mask, dim, dtype), torch.nn.functional.masked_softmax(input, mask, dim, dtype))

def torch.nn.functional.masked_softmax(self):
    self._test_masked_softmax(torch.ones((2, 2), ....)
```

File: torch/nn/functional.py [hunk 4]
```diff
@@ -4037,6 +4037,27 @@ def _pad_circular(input, padding):
...
+
+    # Input assumed to be float32. Type checking:
+    if input.dtype != torch.float32:
+        raise TypeError('input must be of type torch.float32')
+
+    # To follow the logic of multi_head_attention_forward below,
+    # this assumes the mask will be True where we want to zero out the
+    # entries, and False where we want to preserve them.
+    input = input.masked_fill(mask, float('-inf'))
+    return torch.nn.functional.softmax(input, dim=dim, dtype=dtype)
```
Comment (cpuhrsch) [inline torch/nn/functional.py:4060]: You could just call into C++ here and store the reference within the tests only

File: aten/src/ATen/native/native_functions.yaml [hunk 5]
```diff
@@ -3153,6 +3153,23 @@
     CPU: softmax_backward_cpu
     CUDA: softmax_backward_cuda
 
+- func: masked_softmax.int(Tensor self, Tensor mask, int dim, ScalarType? dtype=None) -> Tensor
```
Comment (gchanan) [inline aten/src/ATen/native/native_functions.yaml:3156]: nit: can we put both of the non-underscore versions under `python_module: nn` (see other examples in this file).  You'll also have to change the implementation in nn/functional.py to call the updated function, eg. as in https://github.com/pytorch/pytorch/blob/74d6a6106cf7e0629886b23594fb81c0b8e434ab/torch/nn/functional.py#L459.
